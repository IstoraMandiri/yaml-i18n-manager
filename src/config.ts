import yaml from 'yaml';
import { promises } from 'fs';
import { resolve } from 'path';
import { log } from './utils';

const { writeFile, readFile, mkdir } = promises;

export const defaultConfig = {
  defaultLocale: 'en',
  contentDir: 'content',
  yamlDir: '_i18n',
  csvDir: 'csv',
  filterKeys: ['link', 'className', 'date', 'tags'],
  filterMarkdown: [],
};

export async function initConfig(argv: any) {
  const [command] = argv._;
  const location = resolve(process.env.PWD, argv.config);
  if (command === 'init') {
    try {
      const path = location.split('/').slice(0, -1).join('/');
      await mkdir(path, { recursive: true });
      await writeFile(location, yaml.stringify(defaultConfig));
      log(`Wrote config to ${argv.config}`, defaultConfig);
      return null;
    } catch (e) {
      throw e;
    }
  }
  let configData, config;
  try {
    configData = await readFile(location, 'utf8');
  } catch (e) {
    log(location);
    log(
      'Config file not found. Pass a config file with `--config` or initialize with `init` command.',
    );
    return null;
  }
  try {
    config = yaml.parse(configData);
  } catch (e) {
    log(`Could not parse the config file at ${location}`);
    return null;
  }
  return config;
}
