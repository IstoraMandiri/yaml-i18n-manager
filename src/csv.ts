import csv from 'async-csv';

const columns = [
  { key: 'dHash', header: 'dHash' },
  { key: 'path', header: 'Location' },
  { key: 'sig', header: 'Key' },
  { key: 'defaultValue', header: 'Default Language' },
  { key: 'value', header: 'Existing Translation' },
  { key: 'update', header: 'Updated Translation (or leave blank)' },
];

export default async function generate(data: TranslatedYaml[]) {
  return await csv.stringify(data, { header: true, columns });
}
