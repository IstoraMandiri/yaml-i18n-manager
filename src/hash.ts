import { createHash } from 'crypto';

export default function (str: string): string {
  return createHash('sha256').update(str, 'utf8').digest('base64').slice(0, 10);
}
