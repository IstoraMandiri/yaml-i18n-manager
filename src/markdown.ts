import yaml from 'yaml';
import fs from 'fs-extra';
import { resolve } from 'path';
import { Config } from './types';
import { scanDir } from './utils';
import { translateYaml } from './translate';

async function parseMarkdonwFile(file: string): Promise<{ header?: any; body: string }> {
  const data = await fs.readFile(file, 'utf8');
  let header, body;
  if (data.startsWith('---')) {
    const [_, _header, _body] = data.split('---');
    body = _body;
    header = yaml.parse(_header);
  } else {
    body = data;
  }
  return { header, body };
}

function normalize(n: string): string {
  const str = n.split('.').slice(0, -2).join('.');
  console.log(str);
  return str;
}

export async function generateMarkdown(locale: string, opts: Config) {
  const markdownFiles = (await scanDir(opts.contentDir))
    .filter(({ ext }) => ['md', 'mdx'].includes(ext))
    .filter(({ relative }) => !opts.filterMarkdown.find((p) => relative.startsWith(p)))
    .map((md) => ({ ...md, locale: md.name.split('.').slice(-2)[0] }))
    .filter(({ locale }) => [opts.defaultLocale, locale].includes(locale));
  const deduped = markdownFiles.filter(
    ({ relative, locale }) =>
      locale === opts.defaultLocale &&
      !markdownFiles.find(
        ({ relative: r2, locale: l2 }) => l2 !== locale && normalize(r2) === normalize(relative),
      ),
  );
  const contents = await Promise.all(
    deduped.map(async (file) => ({ ...file, data: await parseMarkdonwFile(file.path) })),
  );
  const normalized = contents.reduce((a, i): any => {
    const {
      data: { header, body },
      relative,
    } = i;
    const items = [{ relative, key: 'body', value: body }];
    Object.keys(header)
      .filter((k) => !opts.filterKeys.includes(k))
      .forEach((k) => {
        items.push({ relative, key: `header.${k}`, value: header[k] });
      });
    return [...a, ...items];
  }, []);

  const translated = await translateYaml(normalized, opts.defaultLocale, locale);
  const constructed = translated.reduce((o: any, { relative, key, value }: any) => {
    const [k, subKey] = key.split('.');
    const update = o[relative] || {};
    update[k] = subKey ? { ...update[k], [subKey]: value } : value;
    return { ...o, [relative]: update };
  }, {});
  // now write the files.
  await Promise.all(
    Object.keys(constructed).map(async (k) => {
      const newFileName = `${k.split('.').slice(0, -2).join('.')}.${locale}.${k.split('.').pop()}`;
      const { header, body } = constructed[k];
      const data = `---\n${yaml.stringify(header).trim()}\n---${body}`;
      await fs.writeFile(resolve(opts.contentDir, newFileName), data);
    }),
  );
}
