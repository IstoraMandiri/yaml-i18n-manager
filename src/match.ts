import { normalizeSig } from './utils';

export default function match(yamls: ExistingYaml[]): ExistingYaml[] {
  return yamls.map((dYaml) => {
    const matches = yamls.reduce((o, tYaml) => {
      if (tYaml.locale == dYaml.locale || normalizeSig(tYaml) !== normalizeSig(dYaml)) {
        return o;
      }
      return { [tYaml.locale]: tYaml };
    }, {});
    return { ...dYaml, matches };
  });
}
