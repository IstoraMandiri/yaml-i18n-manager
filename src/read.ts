import YAML from 'yaml';
import { resolve, relative } from 'path';
import { promises } from 'fs';
import { hash, K_SEP } from './utils';
const { readdir, readFile } = promises;

import type { ExistingYaml } from './types';

interface ParsedFile {
  absolutePath: string;
  relativePath: string;
  yaml: any[];
}

async function parseFile(targetDir: string, absolutePath: string): Promise<ParsedFile> {
  const data = await readFile(absolutePath, 'utf8');
  const yaml = await YAML.parse(data);
  return {
    absolutePath,
    relativePath: relative(targetDir, absolutePath),
    yaml,
  };
}

async function getFiles(root: string, dir?: string): Promise<ParsedFile[]> {
  const target = dir || root;
  const dirEntries = await readdir(target, { withFileTypes: true });
  const files = await Promise.all(
    dirEntries.map((entry): any => {
      const res = resolve(target, entry.name);
      if (entry.isDirectory()) {
        return getFiles(root, res);
      }
      if (entry.name.endsWith('.yaml') && !entry.name.includes('.collection.')) {
        return parseFile(root, res);
      }
    }),
  );
  return Array.prototype.concat(...files.filter((i) => i));
}

function flatten<T extends Record<string, any>>(
  object: T,
  path: string | null = null,
  separator = '__',
): T {
  return Object.keys(object).reduce((acc: T, key: string): T => {
    // always ignore `key` key
    // ignore non-strings
    if (key === 'key') {
      return acc;
    }
    let newPath = [path, key].filter(Boolean).join(separator);
    const i = key;
    const k = object?.[key].key;
    if (k) {
      newPath = [path, `[${i}:${k}]`].filter(Boolean).join(separator);
    }
    return typeof object?.[key] === 'object'
      ? { ...acc, ...flatten(object[key], newPath, separator) }
      : // filter non-strings
      typeof object[key] !== 'string'
      ? acc
      : { ...acc, [newPath]: object[key] };
  }, {} as T);
}

function normalize(files: ParsedFile[]): ExistingYaml[] {
  return files.reduce((a1, { yaml, relativePath: path }): any => {
    const [locale] = path.split('.').slice(-2);
    const flat = flatten(yaml) as { [k: string]: any };
    return [
      ...a1,
      ...Object.keys(flat).reduce(
        (a2, sig): any => [
          ...a2,
          {
            path,
            locale,
            sig,
            key: sig.split(K_SEP).slice(-1)[0],
            value: flat[sig],
            vHash: hash(flat[sig]),
          },
        ],
        [],
      ),
    ];
  }, []);
}

export default async function read(targetDir: string) {
  const files = await getFiles(targetDir);
  return normalize(files);
}
