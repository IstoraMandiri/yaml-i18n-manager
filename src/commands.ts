import read from './read';
import translate from './translate';
import { combine, match, log } from './utils';
import { parse, generate as generateCsv } from './csv';
import eject from './eject';

import type { Config } from './types';
import generateYaml from './yaml';

async function getTranslations(targetLocale: string, opts: Config) {
  const { contentDir, filterKeys, defaultLocale } = opts;
  const existing = await read(opts);
  const filtered = existing.filter(({ key, value }) => {
    return filterKeys.indexOf(key) === -1 && !value.startsWith('http');
  });
  const matched = match(filtered).filter(({ locale }) => locale === defaultLocale);
  const requiresTranslating = matched.filter(({ matches }) => !matches[targetLocale]);
  const translated = await translate(requiresTranslating, defaultLocale, targetLocale);
  return combine(matched, translated, targetLocale);
}

const commands: any = {
  generate: async ({ locale }: { locale: string }, config: Config) => {
    // generate
    const translations = await getTranslations(locale, config);
    const outputDir = `${config.outputDir || config.contentDir}/${config.yamlDir}`;
    const filePath = `${outputDir}/${locale}.yaml`;
    await generateYaml(translations, filePath);
    log(`Wrote ${translations.length} translations to ${filePath}`);
  },
  eject: async ({ locale }: { locale: string }, config: Config) => {
    const translations = await getTranslations(locale, config);
    const outputDir = config.outputDir || config.contentDir;
    await eject(translations, outputDir);
    log(`Wrote ${translations.length} translations to ${outputDir}`);
  },
  export: async ({ locale }: { locale: string }, config: Config) => {
    const translations = await getTranslations(locale, config);
    const filePath = `${config.csvDir}/${locale}.csv`;
    await generateCsv(translations, filePath);
    log(`Wrote to ${filePath}`);
  },
  import: async ({ csv }: { csv: string }, config: Config) => {
    const imports = await parse(csv);
    const outputDir = config.outputDir || config.contentDir;
    await eject(imports, outputDir);
    log(`Wrote ${imports.length} imports to ${outputDir}`);
  },
};

export default commands;
