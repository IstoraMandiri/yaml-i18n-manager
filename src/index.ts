#!/usr/bin/env node
import yargs from 'yargs';
import { defaultConfig, initConfig } from './config';
import commands from './commands';

function requireLocale(y: any) {
  y.positional('locale', {
    describe: 'Target language code (e.g. ja)',
    type: 'string',
  });
}

const argv: any = yargs
  .usage('Usage: $0 <command> [options]')
  .version('v')
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help')
  .options({
    c: {
      alias: 'config',
      type: 'string',
      describe: 'Config file',
      default: `${defaultConfig.managementDir}/config.yaml`,
    },
  })
  .command('init', 'Generate Config File')
  .command('generate <locale>', 'Auto-translate content', requireLocale)
  .command('export <locale>', 'Export CSV for a locale', requireLocale)
  .command('import <csv>', 'Import CSV translations', (y: any) => {
    y.positional('csv', {
      describe: 'Relative CSV file path',
      type: 'string',
    });
  })
  .demandCommand(1, 'Please enter a command').argv;

(async () => {
  const config = await initConfig(argv);
  if (!config) {
    return;
  }
  const [command] = argv._;
  await commands[command](argv, config);
})();
