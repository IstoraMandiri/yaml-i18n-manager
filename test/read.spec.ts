import read from '../src/read';
import { defaultConfig } from '../src/config';

const target = './test/data/content-input';

const config = { ...defaultConfig, contentDir: target };

describe('read', () => {
  it('returns the key/values', async () => {
    expect(await read(config)).toMatchSnapshot();
  });
});
