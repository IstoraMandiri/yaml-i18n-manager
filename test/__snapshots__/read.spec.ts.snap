// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read returns the key/values 1`] = `
Array [
  Object {
    "fullSig": "menu/home__title",
    "key": "title",
    "locale": "en",
    "path": "menu.en.yaml",
    "sig": "home__title",
    "vHash": "OnhpU4izi1",
    "value": "Home",
  },
  Object {
    "fullSig": "menu/home__link",
    "key": "link",
    "locale": "en",
    "path": "menu.en.yaml",
    "sig": "home__link",
    "vHash": "il7asoJjJE",
    "value": "/",
  },
  Object {
    "fullSig": "menu/about__title",
    "key": "title",
    "locale": "en",
    "path": "menu.en.yaml",
    "sig": "about__title",
    "vHash": "Tvyg0Qxf64",
    "value": "About",
  },
  Object {
    "fullSig": "menu/about__link",
    "key": "link",
    "locale": "en",
    "path": "menu.en.yaml",
    "sig": "about__link",
    "vHash": "l5vdxKjKr9",
    "value": "/about",
  },
  Object {
    "fullSig": "menu/news__title",
    "key": "title",
    "locale": "en",
    "path": "menu.en.yaml",
    "sig": "news__title",
    "vHash": "aXUvIxOy8B",
    "value": "News",
  },
  Object {
    "fullSig": "menu/news__link",
    "key": "link",
    "locale": "en",
    "path": "menu.en.yaml",
    "sig": "news__link",
    "vHash": "9GsW7VpO30",
    "value": "/news",
  },
  Object {
    "fullSig": "sub/welcome/intro",
    "key": "intro",
    "locale": "en",
    "path": "sub/welcome.en.yaml",
    "sig": "intro",
    "vHash": "O7zuwuv2p5",
    "value": "Welcome to this sub folder",
  },
  Object {
    "fullSig": "sub/welcome/description",
    "key": "description",
    "locale": "en",
    "path": "sub/welcome.en.yaml",
    "sig": "description",
    "vHash": "luDxWK7tmC",
    "value": "This is a description that won't be translated",
  },
  Object {
    "fullSig": "sub/welcome/items__test__title",
    "key": "title",
    "locale": "en",
    "path": "sub/welcome.en.yaml",
    "sig": "items__test__title",
    "vHash": "6Aaikc/D5h",
    "value": "Testing",
  },
  Object {
    "fullSig": "sub/welcome/items__hello__title",
    "key": "title",
    "locale": "en",
    "path": "sub/welcome.en.yaml",
    "sig": "items__hello__title",
    "vHash": "UJJQgXm1Ai",
    "value": "Hello Work",
  },
  Object {
    "fullSig": "ui.global/title",
    "key": "title",
    "locale": "en",
    "path": "ui.global.en.yaml",
    "sig": "title",
    "vHash": "fozSBW2nOn",
    "value": "Title",
  },
  Object {
    "fullSig": "ui.global/description",
    "key": "description",
    "locale": "en",
    "path": "ui.global.en.yaml",
    "sig": "description",
    "vHash": "DdECGc15NC",
    "value": "This is a description",
  },
  Object {
    "fullSig": "ui.global/new",
    "key": "new",
    "locale": "en",
    "path": "ui.global.en.yaml",
    "sig": "new",
    "vHash": "V7FjiOgcRD",
    "value": "Some new key",
  },
  Object {
    "fullSig": "ui.global/body",
    "key": "body",
    "locale": "en",
    "path": "ui.global.en.yaml",
    "sig": "body",
    "vHash": "zcRMuPQjiV",
    "value": "## This is some markdown

How is this?
",
  },
  Object {
    "fullSig": "ui.global/footer",
    "key": "footer",
    "locale": "ja",
    "path": "ui.global.ja.yaml",
    "sig": "footer",
    "vHash": "nLFiSv59Yu",
    "value": "Footer here
",
  },
  Object {
    "fullSig": "menu/hello",
    "key": "hello",
    "locale": "ja",
    "path": "menu.ja.yaml",
    "sig": "hello",
    "vHash": "FW0ofrgfwt",
    "value": "This is a translated key
",
  },
  Object {
    "fullSig": "sub/welcome/items__test__title",
    "key": "title",
    "locale": "ja",
    "path": "sub/welcome.ja.yaml",
    "sig": "items__test__title",
    "vHash": "XOdEpb8l5T",
    "value": "This is the title!
",
  },
]
`;
